builder/chroots.go

```diff
@@ -80,6 +82,416 @@ func readBuildChrootsConfig(path string) (*buildChrootsConfig, error) {
+// prefix. Short term, resolve file paths using the symlinks from the
+// filesystem. Long term, we need to check every path component and resolve.
+func resolveFileName(path string) string {
+	re := regexp.MustCompile(`^(/bin/)`)
+	path = re.ReplaceAllLiteralString(path, "/usr/bin/")
+	if strings.HasPrefix(path, "/usr/bin/") {
+		return path
+	}
+	re = regexp.MustCompile(`^(/usr/sbin/)`)
+	path = re.ReplaceAllLiteralString(path, "/usr/bin/")
+	if strings.HasPrefix(path, "/usr/bin/") {
+		return path
+	}
+	re = regexp.MustCompile(`^(/sbin/)`)
+	path = re.ReplaceAllLiteralString(path, "/usr/bin/")
+	if strings.HasPrefix(path, "/usr/bin/") {
+		return path
+	}
+	re = regexp.MustCompile(`^(/lib64/)`)
+	path = re.ReplaceAllLiteralString(path, "/usr/lib64/")
+	if strings.HasPrefix(path, "/usr/lib64/") {
+		return path
+	}
+	re = regexp.MustCompile(`^(/lib/)`)
+	return re.ReplaceAllLiteralString(path, "/usr/lib64/")
+}
+
+func addFileAndPath(destination map[string]bool, absPathsToFiles ...string) {
+	for _, file := range absPathsToFiles {
+		if isBannedPath(file) {
+			continue
+		}
+
+		path := "/"
+		dir := filepath.Dir(file)
+		// skip first empty string so we don't add '/'
+		for _, part := range strings.Split(dir, "/")[1:] {
+			path = filepath.Join(path, part)
+			destination[path] = true
+		}
+		destination[file] = true
+	}
+}
+
+func addOsCoreSpecialFiles(bundle *bundle) {
+	filesToAdd := []string{
+		"/usr/lib/os-release",
+		"/usr/share/clear/version",
+		"/usr/share/clear/versionstamp",
+	}
+
+	addFileAndPath(bundle.Files, filesToAdd...)
+}
+
+func addUpdateBundleSpecialFiles(b *Builder, bundle *bundle) {
+	filesToAdd := []string{
+		"/usr/share/defaults/swupd/contenturl",
+		"/usr/share/defaults/swupd/versionurl",
+		"/usr/share/defaults/swupd/format",
+	}
+
+	if _, err := os.Stat(b.Cert); err == nil {
+		filesToAdd = append(filesToAdd, "/usr/share/clear/update-ca/Swupd_Root.pem")
+	}
+
+	addFileAndPath(bundle.Files, filesToAdd...)
+}
+
+func resolveFilesForBundle(bundle *bundle, packagerCmd []string) error {
+	bundle.Files = make(map[string]bool)
+	for p := range bundle.DirectPackages {
+		queryString := merge(packagerCmd, "repoquery", "-l", "--quiet", p)
+		outBuf, err := helpers.RunCommandOutput(false, queryString[0], queryString[1:]...)
+		if err != nil {
+			return err
+		}
+		for _, f := range strings.Split(outBuf.String(), "\n") {
+			if len(f) > 0 {
+				addFileAndPath(bundle.Files, resolveFileName(f))
+			}
+		}
+	}
+	addFileAndPath(bundle.Files, fmt.Sprintf("/usr/share/clear/bundles/%s", bundle.Name))
+	fmt.Printf("Bundle %s\t%d files\n", bundle.Name, len(bundle.Files))
+	return nil
+}
+
+func resolveFiles(numWorkers int, set bundleSet, packagerCmd []string) error {
+	var err error
+	var wg sync.WaitGroup
+	fmt.Printf("Resolving files using %d workers\n", numWorkers)
+	wg.Add(numWorkers)
+	bundleCh := make(chan *bundle)
+	errorCh := make(chan error, numWorkers)
```

Comment:
I think this code needs much more in the way of explanation. As far as I can see the ideas are very confused. You start up N workers, which keep reading work from the input channel, if they have any error then they write to the error channel, and the parent stops writing to the input chan.

So why does the worker try to read more work after it has sent an error? In particular this can lead to deadlock if you get many errors.

Because you have the error channel buffered this will potentially leave worker routines hanging around as the worker can send an error response and go back to reading the next task, the driver can choose at random if it should read something from the error channel or write something to the task channel, so the worker which has just detected a problem could (at random) start up new work, which can then error.

If the error channel say gets to 90% full, and then the driver exhausts the set and so calls wg.Done, and the remaining workers report more than another 10% of errors, then the workers will hang waiting for their error to be accepted into the error channel, and the driver will hang waiting for all the workers to exit.

When the worker detects an error send it down the error channel and break out of the range and hence call wg.Done rather than trying to create potentially more problems.

This ensures that each worker sends no more than one error response down the channel, and so the error channel is never unable to take all the messages and so the workers never block.

You should drain all of the errorCh, not just take either 0 or 1 value from it after you call wg.Wait.